mne/preprocessing/artifact_detection.py

```diff
@@ -0,0 +1,316 @@
+                                   chpi_locs_dev, chpi_median_pos)
+
+    return annot, hpi_disp, dev_head_t
+
+
+def compute_average_dev_head_t(raw, pos):
+    '''Get new device to head transform based on averaged recording
+
+    It excludes pos that have a BAD_ annotation'''
+
+    time = pos[:, 0]
+    info = raw['info']
+    quats = pos[:, 1:7]
+
+    # Original head pos from file
+    chpi_locs_dev = sorted([d for d in info['hpi_results'][-1]
+                            ['dig_points']], key=lambda x: x['ident'])
+    chpi_locs_dev = np.array([d['r'] for d in chpi_locs_dev])
+
+    # Get head pos displacement
+    chpi_mov_head = np.array([_apply_quat(quat, chpi_locs_dev, move=True)
+                              for quat in quats])
+
+    # Get weights for each head pos
+    weights = np.append(time[1:] - time[:-1], 0)
+
+    # Bad segments have a weight of 0
+    for annoth in raw.annotations:
+        if 'bad_' in annoth['description']:
+            onset = annoth['onset']
+            end = onset + annoth['duration']
+            out_beg = np.argmin(np.abs(time - onset))
+            out_end = np.argmin(np.abs(time - end))
+            weights[out_beg:out_end] = 0
+    weights /= sum(weights)
+
+    # Get head pos closest to weighted average pos
+    avg_head_pos = np.average(chpi_mov_head, axis=0, weights=weights)
+    hpi_disp = chpi_mov_head - np.tile(avg_head_pos, (len(time), 1, 1))
+    hpi_dist = np.sqrt((hpi_disp.reshape(-1, 9) ** 2).sum(axis=1))
+    head_pos_avg = chpi_mov_head[hpi_dist.argmin(), :, :]
+
+    # Compute new dev->head transformation from median
+    dev_head_t = _quaternion_align(info['dev_head_t']['from'],
+                                   info['dev_head_t']['to'],
+                                   chpi_locs_dev, head_pos_avg)
+
+    return dev_head_t
+
+
+
+def detect_muscle(raw, thr=1.5, t_min=1, notch=[60, 120, 180]):
+    """Find and annotate mucsle artifacts based on high frequency activity.
+
+    Data is band pass filtered at high frequencies, smoothed with the envelope,
+    z-scored, channel averaged and low-pass filtered to remove transient peaks.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The data.
+    thr : integer
+        threshold where a segment is marked as artifactual. The thr represent
+        the channel averaged z-score values of the data band pass filtered
+        betweehn 110 and 140 Hz.
+    t_min = integer
+        Min. time between annotated muscle artifacts, below will be annotated
+        as muscle artifact.
+    notch = List
+        The frequencies at which to apply a notch filter
+
+    Returns
+    -------
+    annot : mne.Annotations
+        periods with muscle artifacts
+    art_scores_filt : array
+        the z-score values of the filtered data
+    """
+    raw.pick_types(meg=True, ref_meg=False)
+    if notch is not None:
+        raw.notch_filter(notch, fir_design='firwin')
+    raw.filter(110, 140, fir_design='firwin')
```

Comment:
Raw is not returned, only annotations and z-scores.

I would not let users change the freq. these are standard

Reply:
You do not return the raw, but you change it in-place, so it will be different for the user even if you don't return it (unless you copy). Freq - ok, it's just sfreq dependent - for low sfreq you may not have much in this freq range. But we could change it when it becomes an issue.

Reply:
Nice catch, I was changing raw!